state/workingset.go

```diff
@@ -0,0 +1,566 @@
+
+		workingCandidates() map[hash.PKHash]*Candidate
+	}
+
+	// workingset implements Workingset interface, tracks pending changes to account/contract in local cache
+	workingset struct {
+		blkHeight        uint64
+		cachedCandidates map[hash.PKHash]*Candidate
+		savedAccount     map[string]*State        // save account state before being modified in this block
+		cachedAccount    map[hash.PKHash]*State   // accounts being modified in this block
+		cachedContract   map[hash.PKHash]Contract // contracts being modified in this block
+		accountTrie      trie.Trie                // global state trie
+		dao              db.CachedKVStore         // the underlying DB for account/contract storage
+	}
+)
+
+// NewWorkingSet creates a new working set
+func NewWorkingSet(kv db.KVStore, root hash.Hash32B) (WorkingSet, error) {
+	ws := &workingset{
+		cachedCandidates: make(map[hash.PKHash]*Candidate),
+		savedAccount:     make(map[string]*State),
+		cachedAccount:    make(map[hash.PKHash]*State),
+		cachedContract:   make(map[hash.PKHash]Contract),
+		dao:              db.NewCachedKVStore(kv),
+	}
+	tr, err := trie.NewTrieSharedDB(ws.dao, trie.AccountKVNameSpace, root)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate state trie from config")
+	}
+	ws.accountTrie = tr
+	if err := ws.accountTrie.Start(context.Background()); err != nil {
+		return nil, errors.Wrapf(err, "failed to load state trie from root = %x", root)
+	}
+	return ws, nil
+}
+
+func (ws *workingset) workingCandidates() map[hash.PKHash]*Candidate {
+	return ws.cachedCandidates
+}
+
+//======================================
+// State/Account functions
+//======================================
+// LoadOrCreateState loads existing or adds a new State with initial balance to the factory
+// addr should be a bech32 properly-encoded string
+func (ws *workingset) LoadOrCreateState(addr string, init uint64) (*State, error) {
+	h, err := iotxaddress.GetPubkeyHash(addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "error when getting the pubkey hash")
+	}
+	addrHash := byteutil.BytesTo20B(h)
+	state, err := ws.cachedState(addrHash)
+	switch {
+	case errors.Cause(err) == ErrAccountNotExist:
+		balance := big.NewInt(0)
+		balance.SetUint64(init)
+		state = &State{
+			Balance:      balance,
+			VotingWeight: big.NewInt(0),
+		}
+		ws.cachedAccount[addrHash] = state
+	case err != nil:
+		return nil, errors.Wrapf(err, "failed to get state of %x from cached state", addrHash)
+	}
+	return state, nil
+}
+
+// Balance returns balance
+func (ws *workingset) balance(addr string) (*big.Int, error) {
+	if saved, ok := ws.savedAccount[addr]; ok {
+		return saved.Balance, nil
+	}
+	pkHash, err := iotxaddress.GetPubkeyHash(addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "error when getting the pubkey hash")
+	}
+	state, err := ws.getState(byteutil.BytesTo20B(pkHash))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get state of %x", pkHash)
+	}
```

Comment:
we may call the private function "state"

Reply:
getState() is a lower-level internal func. there is already a private func "state", which calls it.